core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java

```diff
@@ -249,6 +294,77 @@ public void testCloudSqlRefreshesExpiredData() throws Exception {
         () -> instance.getSslData(TEST_TIMEOUT_MS) == data.getSslData(), 1000L);
   }
 
+  @Test
+  public void testThatForceRefreshBalksWhenAScheduledRefreshIsInProgress() throws Exception {
+    // Set expiration 1 minute in the future, so that it will trigger a scheduled refresh
+    // and won't expire during this testcase.
+    InstanceData aboutToExpireData =
+        new InstanceData(
+            null, new SslData(null, null, null), Instant.now().plus(1, ChronoUnit.MINUTES));
+
+    // Set the next refresh data expiration 1 hour in the future.
+    InstanceData data =
+        new InstanceData(
+            null, new SslData(null, null, null), Instant.now().plus(1, ChronoUnit.HOURS));
+
+    AtomicInteger refreshCount = new AtomicInteger();
+    final PauseCondition refresh0 = new PauseCondition();
+    final PauseCondition refresh1 = new PauseCondition();
+
+    CloudSqlInstance instance =
+        new CloudSqlInstance(
+            "project:region:instance",
+            (instanceName, accessTokenSupplier, authType, executor, keyPair) -> {
+              int c = refreshCount.get();
+              InstanceData refreshResult = data;
+              switch (c) {
+                case 0:
+                  refresh0.pause();
+                  refreshResult = aboutToExpireData;
+                  break;
+                case 1:
+                  refresh1.pause();
+                  break;
+              }
+              refreshCount.incrementAndGet();
+              return Futures.immediateFuture(refreshResult);
+            },
+            AuthType.PASSWORD,
+            stubCredentialFactory,
+            executorService,
+            keyPairFuture,
+            TEST_RATE_LIMITER);
+
+    refresh0.proceed();
+    refresh0.waitForPauseToEnd(1000);
+    refresh0.waitForCondition(() -> refreshCount.get() > 0, 1000);
+    // Get the first data that is about to expire
+    assertThat(refreshCount.get()).isEqualTo(1);
+    SslData d = instance.getSslData(TEST_TIMEOUT_MS);
+    assertThat(d).isSameInstanceAs(aboutToExpireData.getSslData());
+
+    // Because the data is about to expire, scheduled refresh will begin immediately.
+    // Wait until refresh is in progress.
+    refresh1.waitForPauseToStart(1000);
+
+    // Then call forceRefresh(), which should balk because a refresh attempt is in progress.
+    assertThat(instance.forceRefresh()).isFalse();
```

Comment:
Since we're only using the boolean here in test, I think we shouldn't assert this way. Instead two calls to force refresh plus a check on the number of times it was called should be adequate, right?

Reply:
Good point. Fixed.